### PR TITLE
Implement visitor for AST 

### PR DESCRIPTION
**Pull request by zfy0701**
resolve https://github.com/AfterShip/clickhouse-sql-parser/issues/8

### CODE REVIEW

**Comment by zfy0701** [comment]
there are several ways to implement visitor pattern, this is my preference but I'm happy to implement this as your team feels right

**Comment by git-hulk** [comment]
@zfy0701 Thanks for your great contribution.

The current visitor pattern looks good to me. I will take another pass before merging.

**Review by git-hulk — DISMISSED** [review]
LGTM

File: parser/visitor_test.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+func TestVisitor_Identical(t *testing.T) {
+	visitor := DefaultASTVisitor{}
+
+	for _, dir := range []string{"./testdata/dml", "./testdata/ddl", "./testdata/query", "./testdata/basic"} {
+		outputDir := dir + "/format"
+
+		entries, err := os.ReadDir(dir)
+		if err != nil {
+			require.NoError(t, err)
+		}
```
Comment (Lance726) [inline parser/visitor_test.go:23]:
```suggestion
		require.NoError(t, err)
```

File: parser/visitor_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+func TestVisitor_SimpleRewrite(t *testing.T) {
+	visitor := simpleRewriteVisitor{}
+
+	sql := `SELECT a, COUNT(b) FROM group_by_all GROUP BY CUBE(a) WITH CUBE WITH TOTALS ORDER BY a;`
+	parser := NewParser(sql)
+	stmts, err := parser.ParseStatements()
+	require.NoError(t, err)
+
+	require.Equal(t, len(stmts), 1)
```
Comment (Lance726) [inline parser/visitor_test.go:82]:
```suggestion
	require.Equal(t, 1, len(stmts))
```

File: parser/visitor_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+func TestVisitor_NestRewrite(t *testing.T) {
+	visitor := nestedRewriteVisitor{}
+
+	sql := `SELECT replica_name FROM system.ha_replicas UNION DISTINCT SELECT replica_name FROM system.ha_unique_replicas format JSON`
+	parser := NewParser(sql)
+	stmts, err := parser.ParseStatements()
+	require.NoError(t, err)
+
+	require.Equal(t, len(stmts), 1)
```
Comment (Lance726) [inline parser/visitor_test.go:124]:
```suggestion
	require.Equal(t, 1, len(stmts))
```

**Review by Lance726 — COMMENTED** [review]
Just a few minor updates, the rest of LGTM.

**Comment by zfy0701** [comment]
done

**Review by Lance726 — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Pull request merged**